### PR TITLE
[#140, #142] Implement NCFPipeline and NCFInferenceComponent

### LINKED ISSUE #140: Implement NCFPipeline

### LINKED ISSUE #142: Implement NCFInferenceComponent

### PR DESCRIPTION
**Pull request by JanKapala**
Closes #140, closes #142

### CODE REVIEW

File: recommender/engines/ncf/training/model_training_step/model_training_step.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+            example_input = (
+                batch[USERS_IDS].to(self.device),
+                batch[USERS].to(self.device),
+                batch[SERVICES_IDS].to(self.device),
+                batch[SERVICES].to(self.device),
+            )
+
+            self.writer.add_graph(model, example_input)
+
+        ncf_optimizer = Adam(model.parameters(), lr=self.learning_rate)
```
Comment (wujuu) [inline recommender/engines/ncf/training/model_training_step/model_training_step.py:257]: on the other hand, an optimizer  type SHOULD be a parameter

File: recommender/engines/ncf/training/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# pylint: disable=line-too-long, too-few-public-methods, useless-super-delegation
+# pylint: disable=fixme
+
+"""Neural Collaborative Filtering Pipeline"""
+
+from recommender.engine.preprocessing.embedder import (
+    Embedder,
+)  # TODO: import embedder from proper module
+from recommender.engines.base.base_pipeline import BasePipeline
+from recommender.engines.ncf.training.data_extraction_step.data_extraction_step import (
```
Comment (wujuu) [inline recommender/engines/ncf/training/pipeline.py:10]: we could leverage the steps to __init__ in ...ncf.training, the imports are unnecessarily convoluted now

**Review by wujuu — CHANGES_REQUESTED** [review]

File: recommender/api/endpoints/recommendations.py [hunk 3]
```diff
@@ -48,7 +81,7 @@ def post(self):
         json_dict = request.get_json()
         agent = load_engine(json_dict)
         try:
-            services_ids = agent.call(json_dict)
+            services_ids = agent(json_dict)
```
[the remark this answers is not in the document]
Comment (wujuu) [reply recommender/api/endpoints/recommendations.py:84]: or inference ?

File: recommender/engines/ncf/training/data_validation_step/data_validation_step.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    test, valid can be more like: 0.98, 0.01, 0.01, but still there should
+    exist at least one user that have at least 10 ordered services -
+    otherwise such a dataset can't be used for effective training.
+    """
+
+    for entry in data:
+        # An explanation of below `5` constant is in the docstring above.
+        if len(entry[ORDERED_SERVICES]) >= 5:
+            return
+    raise TooFewOrderedServicesError()
```
[the remark this answers is not in the document]
Comment (wujuu) [reply recommender/engines/ncf/training/data_validation_step/data_validation_step.py:43]: Should the celery workflow catch all of the pipeline errors at the highest possible call stack level?

File: recommender/engine/preprocessing/preprocessing.py [hunk 5]
```diff
@@ -200,32 +200,30 @@ def user_and_services_to_tensors(user, services):
     and compose them into tensors ready for inference.
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engine/preprocessing/preprocessing.py:200]: In few places we still are - it is in this state for now because I didn't want to refactor whole project too much. After merging it will be refactored.

File: recommender/engines/base/base_steps.py [hunk 6]
```diff
@@ -4,56 +4,65 @@
...
     @abstractmethod
-    def __call__(self, data=None) -> Tuple[object, dict]:
+    def __call__(self, data=None) -> Tuple[Any, Dict]:
         pass
 
+    def resolve_constant(self, name, default=None):
+        """step's config overrides pipeline config
+        pipeline config overrides default"""
+
+        return self.config.get(name, self.pipeline_config.get(name, default))
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engines/base/base_steps.py:23]: We do not allow.

File: recommender/engines/ncf/inference/ncf_inference_component.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    def __init__(self, K: int) -> None:
+        self.neural_cf_model = None
+        super().__init__(K)
+
+    def _load_models(self) -> None:
+        """It loads model or models needed for recommendations and raise
+        exception if it is not available in the database.
+        """
+
+        self.neural_cf_model = NeuralCollaborativeFilteringModel.load(version=NEURAL_CF)
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engines/ncf/inference/ncf_inference_component.py:33]: It's needed, good point.

File: recommender/engines/ncf/training/pipeline.py (hunk 2 above)
Comment (JanKapala) [reply recommender/engines/ncf/training/pipeline.py:10]: okey, could be easily done even now.

File: recommender/engine/preprocessing/normaliser.py [hunk 8]
```diff
@@ -16,8 +16,8 @@
 from recommender.models import User, Service
 
 
-class Normalizer:
-    """User or Services Normalizer"""
+class Normaliser:
```
[the remark this answers is not in the document]
Comment (wujuu) [reply recommender/engine/preprocessing/normaliser.py:19]: if you could

File: recommender/engines/base/base_steps.py [hunk 9]
```diff
@@ -4,56 +4,65 @@
...
 
 
 class ModelValidationStep(BaseStep):
     def __init__(self, config):
         super().__init__(config)
-        pass
+        self.config = self.pipeline_config[ModelValidationStep.__name__]
 
     @abstractmethod
     def __call__(self, data=None) -> Tuple[object, object]:
```
Comment (wujuu) [inline recommender/engines/base/base_steps.py:68]: unncecessary

Comment (wujuu) [reply recommender/engines/base/base_steps.py:68]: also, some base steps inherit from ABC and some don't

**Review by wujuu — APPROVED** [review]
